RFC-0004-support-distributed-procedure.md

```diff
@@ -0,0 +1,397 @@
+```
+
+Reference:
+* https://iceberg.apache.org/docs/nightly/spark-procedures/#rewrite_data_files
+* https://iceberg.apache.org/docs/nightly/spark-procedures/#rewrite_position_delete_files
+
+#### BigQuery
+
+BigQuery's system procedure `BQ.REFRESH_MATERIALIZED_VIEW` is used to refresh a materialized view. It involves scanning the data from source table, calculating and rewriting to the target materialized view, so need to be executed distributively:
+
+```sql
+CALL BQ.REFRESH_MATERIALIZED_VIEW(view_name);
+```
+
+Reference:
+* https://cloud.google.com/bigquery/docs/reference/system-procedures#bqrefresh_materialized_view
+
+#### Trino
+
+Trino use its own SQL syntax extension `ALTER TABLE EXECUTE` to support these scenarios. For example, it use optimize command to rewrite the content of the specified table so that it is merged into fewer but larger files:
+
+```sql
+ALTER TABLE test_table EXECUTE optimize;
+ALTER TABLE test_table EXECUTE optimize(file_size_threshold => '128MB');
+ALTER TABLE test_partitioned_table EXECUTE optimize WHERE partition_key = 1;
+```
+
+Reference:
+* https://trino.io/docs/current/connector/iceberg.html#alter-table-execute
+
+## Goals
+
+* Expand the ability of procedures, enable them to leverage worker nodes to manipulate table data as well as metadata, which could be called as `distributed procedure`.
+* Expand currently procedure architecture to support the definition, registration, invocation and execution of `distributed procedures`.
+* Provide clarify architecture hierarchy and well defined SPI interface, enable various connectors to easily support distributed procedures.
+* Support distributed procedure for Iceberg connector, and implement a specific distributed procedure: `rewrite_data_files`.
+
+## Proposed Implementation
+
+### Overview
+
+1. Distributed procedure is used for processing of actual data besides metadata for a target table, such as data rewriting, merging, sorting or repartitioning. We need to re-factor and expand `Procedure/ProcedureRegistry` architecture, so that we can:
+   * Define a distributed procedure and register it into procedure registry
+   * Acquire and use procedure registry in `presto-analyzer` and `connectors` module
+
+
+2. Define a new query type `CALL_DISTRIBUTED_PROCEDURE`, and associate the call distributed procedure statement with this type during preparer phase. So that this kind of statements would fall into the path of `SqlQueryExecutionFactory.createQueryExecution(...)`. The generated `SqlQueryExecution` object can utilize the existing distributed querying, writing, and final committing mechanism to achieve distributed execution for procedures.
+
+
+3. Add a new plan node type: `CallDistributedProcedureNode`. During the analyzing and logical planning phase, construct a logical plan with the following shape for `call distributed procedure` statement:
+
+    ```text
+    TableScanNode -> FilterNode -> CallDistributedProcedureNode -> TableFinishNode -> OutputNode
```

Comment:
Not sure I understand correctly, but if we are going to support a general distributed procedure then why does it have to be between tableScan and TableFinish operators?
If its not for a general case but specific to table layouts, then why not just a custom operator implementation?

Reply:
Thanks for the comment. As described in this RFC, compared to existing coordinator-only procedures, these newly expanded kind of procedures which need to be executed distributively involve operations on table's data other than metadata. For example, these procedures can rewrite table data, merge small data files, sort table data, repartition table data etc. So the common action process of them includes reading data from the target table, doing some filtering and translation, writing the translated data to files and submitting the changes in metadata level. So I think the general distributed procedure could be planned to a plan tree like this. (Or adding some other plan nodes like JoinNode/AggregationNode etc. for functional expansion in future)

Reply:
> why not just a custom operator implementation?

After optimization and plan fragmentation, we should be able to shuffle data between stages to utilize the capabilities of the entire cluster.